homeassistant/components/tplink/fan.py

```diff
@@ -0,0 +1,105 @@
+"""Support for TPLink Fan devices."""
+
+import logging
+import math
+from typing import Any
+
+from kasa import Device, Module
+from kasa.interfaces import Fan
+
+from homeassistant.components.fan import FanEntity, FanEntityFeature
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util.percentage import (
+    percentage_to_ranged_value,
+    ranged_value_to_percentage,
+)
+from homeassistant.util.scaling import int_states_in_range
+
+from . import TPLinkConfigEntry
+from .coordinator import TPLinkDataUpdateCoordinator
+from .entity import CoordinatedTPLinkEntity, async_refresh_after
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: TPLinkConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up fans."""
+    data = config_entry.runtime_data
+    parent_coordinator = data.parent_coordinator
+    device = parent_coordinator.device
+    entities: list[CoordinatedTPLinkEntity] = []
+    if Module.Fan in device.modules:
+        entities.append(
+            TPLinkFan(device, parent_coordinator, device.modules[Module.Fan])
+        )
+    entities.extend(
+        TPLinkFan(child, parent_coordinator, child.modules[Module.Fan], parent=device)
+        for child in device.children
+        if Module.Fan in child.modules
+    )
+    async_add_entities(entities)
+
+
+SPEED_RANGE = (1, 4)  # off is not included
```

Comment:
Would be nice if this came from the lib in a future update

Reply:
Yeah, we discussed about this in private and agreed that it'd be fine to access the feature by its ID (which are now de facto a part of our stable API) to fetch the information from it.

Changing this is not necessary for getting this in as there are no other fan devices, so this is just something we should improve on in the future releases.

Reply:
https://github.com/python-kasa/python-kasa/issues/1008 opened and assigned to 0.7.1